### DISCUSSION TITLE
Initial lens model for DESIJ0215-2909

### DISCUSSION THREAD

**Discussion by numjahan** [AstroBridge/BDLensing#173] [Q&A] [answered]
Considering the bright blob at 1 o'clock as an additional source, I added another `SHAPELETS` to the `source_model_list`.

![image](https://github.com/ajshajib/BDLensing/assets/92924875/12aa7b64-9c12-4a9b-9698-527120207e55)

I kept the `center_x` & `center_y` at zero for both `SHAPELETS`.

![image](https://github.com/ajshajib/BDLensing/assets/92924875/85329311-b996-4cbc-a90c-034a59970f4f)

This is the result I got after running PSO. The reduced chi^2 value is **0.9024**. But here, the gamma value was fixed to 1.9. The notebook can be found [here](https://colab.research.google.com/drive/1YGdOimvIx_8kZkarpwAK14OZCb2OHQ_V?usp=sharing).

![image](https://github.com/ajshajib/BDLensing/assets/92924875/f887f03e-dd49-404e-be77-60d048c2a46f)

 
Then, I rerun it by unfixing it, and the residuals became worse. Now, the reduced chi^2 value is **1.5953**. The gamma value I got is 1.4 which is lower than the typical gamma value. The notebook for this case is [here](https://colab.research.google.com/drive/17-N0EtdgC4lNvLid9AceUCiAJpwM_jd-?usp=sharing).

![image](https://github.com/ajshajib/BDLensing/assets/92924875/2a4cd230-1826-439a-b33a-6f7fef335639)

What steps can I take next to improve it?

**Comment by SamWGamgee** [comment]
Can you show me the other plots?

**Reply by numjahan** [reply]
1st case:

![image](https://github.com/ajshajib/BDLensing/assets/92924875/b44f7ec9-28e6-4606-959f-4e7a8fd08d26)

2nd case:

![image](https://github.com/ajshajib/BDLensing/assets/92924875/99b5dc99-5ecd-478f-9849-196f387451ce)

**Comment by ajshajib** [comment]
The model in the second case has the freedom to find the best-fit values from the first case and that low $\chi^2$ value. Although it's not giving you that result, if you ran the second model for a very long time, it would eventually find the right solution. However, we don't want to run the second case model for a long time for practical reasons. So, it's fine to fix the `gamma` parameter during PSO and then let it be a free parameter for MCMC. However, was there some particular reason to choose 1.9 as the fixed value? The value 2 would be a more standard choice and look less arbitrarily picked.

For comparison, can you also post the model and residual plots from one of your previous runs when you kept the upper blob masked and had only one source galaxy in the model?

**Reply by numjahan** [reply]
I ran it again fixing the upper and lower value of gamma. The residual looks like this -
![image](https://github.com/ajshajib/BDLensing/assets/92924875/7ccb7ebb-8e91-4343-95b9-67ef9646a5ad)

Now the gamma value is 1.91. The reduced chi^2 value is 1.11

**Reply by numjahan** [reply]
This is the best result I got when the upper blob was masked and was considering one source galaxy in the model.
![image](https://github.com/ajshajib/BDLensing/assets/92924875/905f180d-6e63-4e77-9b5c-eddb71a6c634)

**Comment by numjahan** [comment]
The initial gamma value was 1.9. There was no particular reason to choose 1.9 as the initial value. Then, I made it fixed by adding `['update_settings', {'lens_add_fixed': [[0, ['gamma']]]}]`. I thought it would make the `gamma` fixed to a standard value. But it appears that it remained fixed at the initial value of 1.9.

@ajshajib Should I proceed to MCMC for the first case or is there any scope for further improvement?

**Reply by ajshajib** [reply]
Yes, I suggest fixing the `gamma` value to 2 during PSO fitting, then allowing it to be free during MCMC. You should proceed with MCMC after making a PR to close the PSO-issue for this system. 👍

**Reply by numjahan** [reply]
Fixing the `gamma` value to 2.0, I reran it. This is what I got -  
![image](https://github.com/ajshajib/BDLensing/assets/92924875/1d95e401-ae0b-4e11-820a-fca3d4464110)

-1.1393523697581565 reduced X^2 of best position
-3167.969264112554 log likelihood
5561 effective number of data points
[{'theta_E': 1.0753986440587493, 'gamma': 2.0, 'e1': -0.2587257292920046, 'e2': -0.03077883686695652, 'center_x': -0.029408820986434565, 'center_y': -0.016563148039490272}, {'gamma1': -0.005622666106580905, 'gamma2': 0.00012953129139976194, 'ra_0': 0, 'dec_0': 0}] lens result
[{'amp': 1, 'R_sersic': 0.0024181142713267795, 'n_sersic': 1.0, 'e1': 0.4704835931545308, 'e2': -0.11407513906769993, 'center_x': 0.021774330446079376, 'center_y': 0.11247658910745775}, {'amp': 1, 'n_max': 10, 'beta': 0.048642312617611164, 'center_x': 0.021774330446079376, 'center_y': 0.11247658910745775}, {'amp': 1, 'n_max': 10, 'beta': 0.16944430302055807, 'center_x': -0.1720880323632249, 'center_y': 0.3840833915458504}] source result
[{'amp': 1, 'R_sersic': 2.0993947140095632, 'n_sersic': 1.0, 'e1': -0.1594078387376601, 'e2': -0.16833394118601427, 'center_x': -0.061671657801875224, 'center_y': 0.03651190632404214}, {'amp': 1, 'R_sersic': 0.8519918256067333, 'n_sersic': 4.0, 'e1': -0.08694153433630444, 'e2': -0.031355060147655725, 'center_x': -0.061671657801875224, 'center_y': 0.03651190632404214}] lens light result
[] point source result
{} special param result
38149.131774902344 time used for  PSO



- The residual is not looking good as in the first case. Is it okay to start MCMC with this one? Should I run it a couple of times to see if it improves?

**Reply by ajshajib** [reply]
The larger $\chi^2$ value is fine. Once you run MCMC with free `gamma`, it will converge to the preferred value giving you the better $\chi^2$ value.

In your latest model, there are strong residuals here:
![251331210-1d95e401-ae0b-4e11-820a-fca3d4464110 (1)](https://github.com/ajshajib/BDLensing/assets/20169048/23407499-349e-4c4a-b5a1-948a1f13a7a5)

Do you know if this part could be the counter-image of the masked elongated object at 1 o'clock? Can you run a test by unmasking that part?

**Reply by numjahan** [reply]
I have run a test by unmasking the part you mentioned. The blue residual in that specific region seems to persist. The reduced chi^2 value turned out to be very good this time.

Another point to mention is that there is apparently no significant residual in the area where the elongated mask was previously placed. Instead, it is trying to reconstruct the elongated light. So, I think it would be best to keep the region unmasked. Could you please share your thoughts on this?

Should I try to mask the part at 7-8 o'clock where we consistently observe a strong residual?


![image](https://github.com/ajshajib/BDLensing/assets/92924875/67c1e43b-b5ef-4d81-ba60-bc7ab2dbcaa2)
-0.992583046714081 reduced X^2 of best position
-2912.2386590591136 log likelihood
5868 effective number of data points
[{'theta_E': 1.059186791967156, 'gamma': 2.0, 'e1': -0.28078720332966683, 'e2': -0.05597619046296158, 'center_x': -0.0217996212068303, 'center_y': 0.017157627517038726}, {'gamma1': -0.0011376978933713647, 'gamma2': -0.018123929568345097, 'ra_0': 0, 'dec_0': 0}] lens result
[{'amp': 1, 'R_sersic': 0.02240430428898386, 'n_sersic': 1.0, 'e1': 0.17669603745152374, 'e2': -0.1882110093416536, 'center_x': -0.15831518030793396, 'center_y': 0.5924564432389736}, {'amp': 1, 'n_max': 10, 'beta': 0.19972567226457033, 'center_x': -0.15831518030793396, 'center_y': 0.5924564432389736}, {'amp': 1, 'n_max': 10, 'beta': 0.03156554508224456, 'center_x': 0.05026763944703301, 'center_y': 0.11401112889091058}] source result
[{'amp': 1, 'R_sersic': 0.17693460547153014, 'n_sersic': 1.0, 'e1': 0.016194515104388955, 'e2': 0.133373063928328, 'center_x': -0.057991302396753985, 'center_y': 0.03928279130857663}, {'amp': 1, 'R_sersic': 1.6476871409509357, 'n_sersic': 4.0, 'e1': -0.12185122039435012, 'e2': -0.08276857839920684, 'center_x': -0.057991302396753985, 'center_y': 0.03928279130857663}] lens light result
[] point source result
{} special param result
11072.123222827911 time used for  PSO

**Reply by ajshajib** [reply]
Please show me the other plot, I would like to see what the lensed arcs look like.

**Reply by numjahan** [reply]
![image](https://github.com/ajshajib/BDLensing/assets/92924875/cd4ef2fd-3ea7-494a-9168-f9051aa6cf6c)

**Reply by ajshajib** [reply, marked as answer]
Ok, thanks! I don't see anything problematic in the model without the mask at the top. One test to check which is a better model between with or without the mask is to look at the p-value. You can't compare the reduced $\chi^2$ directly for a rigorous comparison because the number of fitted pixels are different. The following is a quick code to compare the p-values:

```
from scipy import stats

# model without mask
# -0.992583046714081 reduced X^2 of best position
# -2912.2386590591136 log likelihood
# 5868 effective number of data points
reduced_chi_2 = 0.992583046714081
degree_of_freedom = 5868
print(f'p-value: {1 - stats.chi2.cdf(reduced_chi_2*degree_of_freedom, degree_of_freedom)}')

# model with mask
# -1.1393523697581565 reduced X^2 of best position
# -3167.969264112554 log likelihood
# 5561 effective number of data points
reduced_chi_2 = 1.1393523697581565
degree_of_freedom = 5561
print(f'p-value: {1 - stats.chi2.cdf(reduced_chi_2*degree_of_freedom, degree_of_freedom)}')
```

Output:
```
p-value: 0.6541649271451053
p-value: 9.726663918740996e-13
```

The model with the higher value p-value (without the mask) is a more acceptable model (by a lot in this case).

**Reply by numjahan** [reply]
Okay, thanks a lot!

**Reply by numjahan** [reply]
@ajshajib Here is another case:

![image](https://github.com/ajshajib/BDLensing/assets/92924875/a02976b0-57d2-4b04-a625-d64091ab076b)
-0.941001195020051 reduced X^2 of best position
-2616.4538227532516 log likelihood
5561 effective number of data points
[{'theta_E': 1.048496247149965, 'gamma': 2.0, 'e1': -0.2133381744507264, 'e2': -0.00930671675237403, 'center_x': -0.029093944652893427, 'center_y': 0.030741206030925918}, {'gamma1': 0.019337515241461656, 'gamma2': 0.0021715539492514297, 'ra_0': 0, 'dec_0': 0}] lens result
[{'amp': 1, 'R_sersic': 0.008873848777890702, 'n_sersic': 1.0, 'e1': 0.450158697015663, 'e2': -0.07802807121085752, 'center_x': 0.0470132203224288, 'center_y': 0.15542697521996557}, {'amp': 1, 'n_max': 10, 'beta': 0.021567679902402173, 'center_x': 0.0470132203224288, 'center_y': 0.15542697521996557}, {'amp': 1, 'n_max': 10, 'beta': 0.17363345832335048, 'center_x': -0.18434663831433507, 'center_y': 0.41475375079173465}] source result
[{'amp': 1, 'R_sersic': 0.16634186463663664, 'n_sersic': 1.0, 'e1': -0.03398978440327349, 'e2': 0.15795360352698923, 'center_x': -0.05889489909834958, 'center_y': 0.03996056073326276}, {'amp': 1, 'R_sersic': 1.6079955087143525, 'n_sersic': 4.0, 'e1': -0.11319521477416801, 'e2': -0.07884069198312885, 'center_x': -0.05889489909834958, 'center_y': 0.03996056073326276}] lens light result
[] point source result
{} special param result
26757.434551000595 time used for  PSO
![image](https://github.com/ajshajib/BDLensing/assets/92924875/dd9174e4-d2de-4af8-9f3e-ea5522afaea9)

If I compare the p-value of this with the one without the mask : 
![image](https://github.com/ajshajib/BDLensing/assets/92924875/5cad300f-1ea1-4cf3-bc10-0b31860da032)

The model with mask has a higher p-value. What do you think I should do then?
